### PR TITLE
📖 update KCP proposal with latest changes

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR updates the KCP proposal document with latest changes, so then it will be possible to extend this proposal with the considerations for implementing automatic remediating of unhealthy control plane machines

/assign @vincepri 
/assign @sedefsavas

### CODE REVIEW

File: docs/proposals/20191017-kubeadm-based-control-plane.md [hunk 1]
```diff
@@ -606,4 +616,6 @@ For the purposes of designing upgrades, two existing lifecycle managers were exa
 - [x] 10/17/2019: Initial Creation
 - [x] 11/19/2019: Initial KubeadmControlPlane types added [#1765](https://github.com/kubernetes-sigs/cluster-api/pull/1765)
 - [x] 12/04/2019: Updated References to ErrorMessage/ErrorReason to FailureMessage/FailureReason
-- [] 12/04/2019: Initial stubbed KubeadmControlPlane controller added [#1826](https://github.com/kubernetes-sigs/cluster-api/pull/1826)
+- [x] 12/04/2019: Initial stubbed KubeadmControlPlane controller added [#1826](https://github.com/kubernetes-sigs/cluster-api/pull/1826)
+- [x] 12/04/2019: Initial stubbed KubeadmControlPlane controller added [#1826](https://github.com/kubernetes-sigs/cluster-api/pull/1826)
```
Comment (sedefsavas) [inline docs/proposals/20191017-kubeadm-based-control-plane.md:620]: Same line exists above.

File: docs/proposals/20191017-kubeadm-based-control-plane.md [hunk 2]
```diff
@@ -453,70 +471,62 @@ spec:
...
-
-- Determining if a Machine is "up to date" will be done through the use of a label (controlplane.cluster.x-k8s.io/hash) that is placed on the Machine at creation time. The value of this label is generated by computing the Hash of the KubeadmControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used (when a new configuration is referenced in a new InfrastructureTemplate).
-  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (as in the Acme cluster example above, the image is stored in InfrastructureTemplate.Spec.OSImage.ID).
+- The rollout algorithm is the following:
+  - Find Machines that need rollout
+  - If there is a machine requiring rollout
+    - Scale up control plane creating a machine with the new spec
+    - Scale down control plane by removing one of the machine that needs rollout (the oldest out-of date machine in the failure domain that has the most out-of-date control-plane machines on it).
```
Comment (sedefsavas) [inline docs/proposals/20191017-kubeadm-based-control-plane.md:522]:
```suggestion
    - Scale down control plane by removing one of the machine that needs rollout (the oldest out-of date machine in the failure domain that has the most control-plane machines on it).
```

**Review by sedefsavas — COMMENTED** [review]
Thanks for updating this proposal. Just a few comments.

File: docs/proposals/20191017-kubeadm-based-control-plane.md (hunk 2 above)
Comment (sedefsavas) [reply docs/proposals/20191017-kubeadm-based-control-plane.md:522]: It is picking a failure domain that has the most number of control plane machines and at least 1 outdated machine in it.

Comment (fabriziopandini) [reply docs/proposals/20191017-kubeadm-based-control-plane.md:522]: If I'm not wrong this is considering only out of date machines:

In `upgradeControlPlane`, the we are calling `scaleDownControlPlane` and passing the list of `machinesRequireUpgrade`
https://github.com/kubernetes-sigs/cluster-api/blob/38d8699bb7c42f10b20a36a553161ba85f739fc7/controlplane/kubeadm/controllers/upgrade.go#L98

Such list is the used for `selectMachineForScaleDown`
https://github.com/kubernetes-sigs/cluster-api/blob/494b8b146055390d0fa49910f5ebdf02e83199d5/controlplane/kubeadm/controllers/scale.go#L103

And being the list not empty it drives also the selection of the failure domains to start deleting from

https://github.com/kubernetes-sigs/cluster-api/blob/494b8b146055390d0fa49910f5ebdf02e83199d5/controlplane/kubeadm/controllers/scale.go#L152-L155

File: docs/proposals/20191017-kubeadm-based-control-plane.md [hunk 3]
```diff
@@ -164,76 +164,91 @@ import (
 
 // KubeadmControlPlaneSpec defines the desired state of KubeadmControlPlane.
 type KubeadmControlPlaneSpec struct {
```
Comment (vincepri) [inline docs/proposals/20191017-kubeadm-based-control-plane.md:166]: Can we remove code and instead link it to where it's hosted?

File: docs/proposals/20191017-kubeadm-based-control-plane.md [hunk 4]
```diff
@@ -453,70 +471,62 @@ spec:
...
-    - `upgradeAfter` field is before now
-  - Mark machine as selected for upgrade
-  - Scale up control plane
-  - Mark machine as having a replacement
-  - Scale down control plane by removing a machine that has been marked for removal and has a replacement already created
-
-- Determining if a Machine is "up to date" will be done through the use of a label (controlplane.cluster.x-k8s.io/hash) that is placed on the Machine at creation time. The value of this label is generated by computing the Hash of the KubeadmControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used (when a new configuration is referenced in a new InfrastructureTemplate).
-  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (as in the Acme cluster example above, the image is stored in InfrastructureTemplate.Spec.OSImage.ID).
+- The rollout algorithm is the following:
+  - Find Machines that need rollout
```
Comment (vincepri) [inline docs/proposals/20191017-kubeadm-based-control-plane.md:519]:
```suggestion
  - Find Machines that have an outdated spec
```

File: docs/proposals/20191017-kubeadm-based-control-plane.md (hunk 2 above)
Comment (sedefsavas) [reply docs/proposals/20191017-kubeadm-based-control-plane.md:522]: While deciding on the failure domain, we use all control-plane machine (c.machines):
https://github.com/kubernetes-sigs/cluster-api/blob/494b8b146055390d0fa49910f5ebdf02e83199d5/controlplane/kubeadm/internal/failure_domain.go#L54-L56

Comment (sedefsavas) [reply docs/proposals/20191017-kubeadm-based-control-plane.md:358]: FailureDomainWithMostMachines is actually considering all machines not just `out-of-date` machines. It picks an fd with most machines that has at least 1 out-of-date machine.

Comment (fabriziopandini) [reply docs/proposals/20191017-kubeadm-based-control-plane.md]: done!

**Comment by fabriziopandini** [comment]
Comment addressed and commit squashed

**Review by vincepri — APPROVED** [review]
/approve
/assign @sedefsavas

**Pull request merged**